### PR TITLE
feat(core): view engine - add missing DI features

### PR DESCRIPTION
**Pull request by tbosch**
First 2 commits are from https://github.com/angular/angular/pull/14216

### CODE REVIEW

File: modules/@angular/core/src/view/types.ts [hunk 1]
```diff
@@ -126,6 +127,7 @@ export enum NodeFlags {
   HasComponent = 1 << 9,
   HasContentQuery = 1 << 10,
   HasViewQuery = 1 << 11,
+  LazyProvider = 1 << 15
```
Comment (tbosch) [inline modules/@angular/core/src/view/types.ts]: should be 12

File: modules/@angular/core/src/view/provider.ts [hunk 2]
```diff
@@ -182,40 +194,105 @@ export function checkAndUpdateProviderDynamic(view: ViewData, def: NodeDef, valu
...
+      injectable = factory(...depValues);
+  }
+  return injectable;
+}
+
+export function resolveDep(view: ViewData, requestNodeIndex: number, depDef: DepDef): any {
+  const notFoundValue = depDef.flags & DepFlags.Optional ? null : Injector.THROW_IF_NOT_FOUND;
   const tokenKey = depDef.tokenKey;
+  let skipSelf = depDef.flags & DepFlags.SkipSelf;
+  let elIndex = findElementDef(view, requestNodeIndex).index;
```
Comment (tbosch) [inline modules/@angular/core/src/view/provider.ts]: Maybe pass elIndex and requestorNodeIndex

File: modules/@angular/core/src/view/provider.ts [hunk 3]
```diff
@@ -182,40 +194,105 @@ export function checkAndUpdateProviderDynamic(view: ViewData, def: NodeDef, valu
...
+      }
+      injectable = factory(...depValues);
+  }
+  return injectable;
+}
+
+export function resolveDep(view: ViewData, requestNodeIndex: number, depDef: DepDef): any {
+  const notFoundValue = depDef.flags & DepFlags.Optional ? null : Injector.THROW_IF_NOT_FOUND;
   const tokenKey = depDef.tokenKey;
+  let skipSelf = depDef.flags & DepFlags.SkipSelf;
```
Comment (tbosch) [inline modules/@angular/core/src/view/provider.ts]: remove

File: modules/@angular/core/test/view/provider_spec.ts [hunk 4]
```diff
@@ -50,18 +50,74 @@ function defineTests(config: {directDom: boolean, viewFlags: number}) {
...
         createAndGetRootNodes(compViewDef([
-          elementDef(NodeFlags.None, null, null, 1, 'span'),
-          providerDef(NodeFlags.None, null, 0, SomeService, [])
+          elementDef(NodeFlags.None, null, null, 2, 'span'),
+          directiveDef(NodeFlags.LazyProvider, null, 0, LazyService, []),
+          directiveDef(NodeFlags.None, null, 0, SomeService, [Injector])
+        ]));
+
+        expect(lazy).toBeUndefined();
+        expect(instance.dep.get(LazyService));
```
Comment (tbosch) [inline modules/@angular/core/test/view/provider_spec.ts]: remove `expect`

**Review by vicb — APPROVED** [review]

File: modules/@angular/core/src/view/types.ts (hunk 1 above)
Comment (tbosch) [inline modules/@angular/core/src/view/types.ts]: Done

File: modules/@angular/core/src/view/provider.ts (hunk 3 above)
Comment (tbosch) [inline modules/@angular/core/src/view/provider.ts]: Done

File: modules/@angular/core/test/view/provider_spec.ts (hunk 4 above)
Comment (tbosch) [inline modules/@angular/core/test/view/provider_spec.ts]: Done

File: modules/@angular/core/src/view/provider.ts (hunk 2 above)
Comment (tbosch) [inline modules/@angular/core/src/view/provider.ts]: Done

**Pull request closed without merging**